src/main/java/uk/gov/justice/digital/delius/service/OffenderPrisonerService.java

```diff
@@ -0,0 +1,65 @@
+package uk.gov.justice.digital.delius.service;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.justice.digital.delius.jpa.standard.entity.*;
+import uk.gov.justice.digital.delius.jpa.standard.repository.EventRepository;
+import uk.gov.justice.digital.delius.jpa.standard.repository.OffenderPrisonerRepository;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+public class OffenderPrisonerService {
+    private final EventRepository eventRepository;
+    private final OffenderPrisonerRepository offenderPrisonerRepository;
+
+    public OffenderPrisonerService(EventRepository eventRepository, OffenderPrisonerRepository offenderPrisonerRepository) {
+        this.eventRepository = eventRepository;
+        this.offenderPrisonerRepository = offenderPrisonerRepository;
+    }
+
+    @Transactional
+    public Offender refreshOffenderPrisonersFor(Offender offender) {
+        final var offenderId = offender.getOffenderId();
+        final var events = eventRepository.findByOffenderId(offenderId);
+        offenderPrisonerRepository.deleteAllByOffenderId(offenderId);
+        offenderPrisonerRepository.saveAll(createOffenderPrisonersFromEvents(offenderId, events));
+        offender.setMostRecentPrisonerNumber(getPrisonNumberFromLatestEvent(events));
+        return offender;
+    }
+
+    public String getPrisonNumberFromLatestEvent(List<Event> events) {
```

Comment:
Just wondering why not `Optional<String>`?

Reply:
we have to set `offender.setMostRecentPrisonerNumber()` if we have a prisoner number or not, or to but it another way if there was no value we would clear it by setting it to blank hence the `orElse("")`